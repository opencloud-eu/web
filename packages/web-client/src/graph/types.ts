import type { AxiosInstance, ResponseType } from 'axios'
import type { Configuration } from './generated'

export interface GraphFactoryOptions {
  axiosClient: AxiosInstance
  config: Configuration
}

export interface GraphRequestOptions {
  headers?: Record<string, string>
  params?: Record<string, string>
  signal?: AbortSignal
  responseType?: ResponseType
}
