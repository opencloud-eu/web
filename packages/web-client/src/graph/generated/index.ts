/* tslint:disable */
/* eslint-disable */
/**
 * Libre Graph API
 * Libre Graph is a free API for cloud collaboration inspired by the MS Graph API.
 *
 * The version of the OpenAPI document: v1.0.7
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

