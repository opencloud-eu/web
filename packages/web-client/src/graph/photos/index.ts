export * from './photo'
export * from './types'
