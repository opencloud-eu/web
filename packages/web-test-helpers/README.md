# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
