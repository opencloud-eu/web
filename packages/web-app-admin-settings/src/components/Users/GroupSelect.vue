<template>
  <div id="user-group-select-form">
    <oc-select
      :model-value="selectedOptions"
      class="oc-mb-s"
      :multiple="true"
      :options="groupOptions"
      option-label="displayName"
      :label="$gettext('Groups')"
      :fix-message-line="true"
      v-bind="$attrs"
      :required-mark="requiredMark"
      @update:model-value="onUpdate"
    >
      <template #selected-option="{ displayName, id }">
        <span class="oc-flex oc-flex-center">
          <avatar-image
            class="oc-flex oc-align-self-center oc-mr-s"
            :width="16.8"
            :userid="id"
            :user-name="displayName"
          />
          <span>{{ displayName }}</span>
        </span>
      </template>
      <template #option="{ displayName, id }">
        <div class="oc-flex">
          <span class="oc-flex oc-flex-center">
            <avatar-image
              class="oc-flex oc-align-self-center oc-mr-s"
              :width="16.8"
              :userid="id"
              :user-name="displayName"
            />
            <span>{{ displayName }}</span>
          </span>
        </div>
      </template>
    </oc-select>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, PropType, ref, unref, watch } from 'vue'
import { Group } from '@opencloud-eu/web-client/graph/generated'

export default defineComponent({
  name: 'GroupSelect',
  props: {
    selectedGroups: {
      type: Array as PropType<Group[]>,
      required: true
    },
    groupOptions: {
      type: Array as PropType<Group[]>,
      required: true
    },
    requiredMark: {
      type: Boolean,
      default: false,
      required: false
    }
  },
  emits: ['selectedOptionChange'],
  setup(props, { emit }) {
    const selectedOptions = ref()
    const onUpdate = (group: Group) => {
      selectedOptions.value = group
      emit('selectedOptionChange', unref(selectedOptions))
    }

    const currentGroups = computed(() => props.selectedGroups)
    watch(
      currentGroups,
      () => {
        selectedOptions.value = props.selectedGroups
          .map((g) => ({
            ...g,
            readonly: g.groupTypes?.includes('ReadOnly')
          }))
          .sort((a: any, b: any) => b.readonly - a.readonly)
      },
      { immediate: true }
    )

    return { selectedOptions, onUpdate }
  }
})
</script>
