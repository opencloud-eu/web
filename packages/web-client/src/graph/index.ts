import { AxiosInstance } from 'axios'
import { Configuration } from './generated'
import { type GraphUsers, UsersFactory } from './users'
import { type GraphGroups, GroupsFactory } from './groups'
import { ApplicationsFactory, GraphApplications } from './applications'
import { DrivesFactory, GraphDrives } from './drives'
import { DriveItemsFactory, GraphDriveItems } from './driveItems'
import { TagsFactory, GraphTags } from './tags'
import { ActivitiesFactory, GraphActivities } from './activities'
import { PermissionsFactory, GraphPermissions } from './permissions'
import { PhotoFactory, GraphPhotos } from './photos'

export interface Graph {
  activities: GraphActivities
  applications: GraphApplications
  tags: GraphTags
  drives: GraphDrives
  driveItems: GraphDriveItems
  users: GraphUsers
  groups: GraphGroups
  permissions: GraphPermissions
  photos: GraphPhotos
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  return <Graph>{
    activities: ActivitiesFactory({ axiosClient, config }),
    applications: ApplicationsFactory({ axiosClient, config }),
    tags: TagsFactory({ axiosClient, config }),
    drives: DrivesFactory({ axiosClient, config }),
    driveItems: DriveItemsFactory({ axiosClient, config }),
    users: UsersFactory({ axiosClient, config }),
    groups: GroupsFactory({ axiosClient, config }),
    permissions: PermissionsFactory({ axiosClient, config }),
    photos: PhotoFactory({ axiosClient, config })
  }
}
