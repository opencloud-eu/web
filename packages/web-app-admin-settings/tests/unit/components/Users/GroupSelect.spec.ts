import GroupSelect from '../../../../src/components/Users/GroupSelect.vue'
import { defaultPlugins, shallowMount } from '@opencloud-eu/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { Group } from '@opencloud-eu/web-client/graph/generated'

const groupMock = mock<Group>({ id: '1', groupTypes: [] })

describe('GroupSelect', () => {
  it('renders the select input', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('correctly maps the read-only state', () => {
    const groupMock = mock<Group>({ id: '1', groupTypes: ['ReadOnly'] })
    const { wrapper } = getWrapper(groupMock)
    expect(wrapper.vm.selectedOptions[0].readonly).toBeTruthy()
  })
  it('emits "selectedOptionChange" on update', () => {
    const group = mock<Group>({ id: '2', groupTypes: [] })
    const { wrapper } = getWrapper()
    wrapper.vm.onUpdate(group)
    expect(wrapper.emitted().selectedOptionChange).toBeTruthy()
    expect(wrapper.vm.selectedOptions).toEqual(group)
  })
})

function getWrapper(group = groupMock) {
  return {
    wrapper: shallowMount(GroupSelect, {
      props: {
        selectedGroups: [group],
        groupOptions: [group]
      },
      global: {
        plugins: [...defaultPlugins()]
      }
    })
  }
}
