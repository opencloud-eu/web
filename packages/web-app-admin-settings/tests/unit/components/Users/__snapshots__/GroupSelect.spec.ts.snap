// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GroupSelect > renders the select input 1`] = `
"<div id="user-group-select-form">
  <oc-select-stub label="Groups" id="oc-select-1" filter="[Function]" disabled="false" labelhidden="false" optionlabel="displayName" searchable="true" clearable="false" loading="false" fixmessageline="true" multiple="true" readonly="false" positionfixed="false" requiredmark="false" model-value="[object Object]" class="mb-2" options="undefined"></oc-select-stub>
</div>"
`;
